tower/src/limit/concurrency/layer.rs

```diff
@@ -1,24 +1,67 @@
+use std::sync::Arc;
+
 use super::ConcurrencyLimit;
+use tokio::sync::Semaphore;
 use tower_layer::Layer;
 
 /// Enforces a limit on the concurrent number of requests the underlying
 /// service can handle.
 #[derive(Debug, Clone)]
-pub struct ConcurrencyLimitLayer {
-    max: usize,
+pub struct ConcurrencyLimitLayer<T = usize> {
+    semaphore: T,
 }
 
 impl ConcurrencyLimitLayer {
     /// Create a new concurrency limit layer.
     pub fn new(max: usize) -> Self {
-        ConcurrencyLimitLayer { max }
+        ConcurrencyLimitLayer { semaphore: max }
+    }
+}
+
+impl<T> ConcurrencyLimitLayer<T>
+where
+    T: ToOwnedSemaphore,
+{
+    /// Create a new concurrency limit layer from a type that implements [`ToOwnedSemaphore`]
+    pub fn new_with_semaphore(semaphore: T) -> Self {
+        ConcurrencyLimitLayer { semaphore }
     }
 }
 
-impl<S> Layer<S> for ConcurrencyLimitLayer {
+impl<S, T> Layer<S> for ConcurrencyLimitLayer<T>
+where
+    T: ToOwnedSemaphore,
+{
     type Service = ConcurrencyLimit<S>;
 
     fn layer(&self, service: S) -> Self::Service {
-        ConcurrencyLimit::new(service, self.max)
+        ConcurrencyLimit::new(service, &self.semaphore)
+    }
+}
+
+/// Produces an owned [`Semaphore`] (`Arc<Semaphore>`)
```

Comment:
I think if we're going to have this trait be part of the API, it may be worth discussing what it does, and how users would use it to create a global concurrency limit across all instances of a stack?